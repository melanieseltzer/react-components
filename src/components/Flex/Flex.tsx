import React, { ReactNode } from 'react';
import styled from 'styled-components';

export interface Flex {
  children: ReactNode;
  background?: string;
  direction?: 'row' | 'row-reverse' | 'column' | 'column-reverse';
  wrap?: 'nowrap' | 'wrap' | 'wrap-reverse';
  justifyContent?:
    | 'flex-start '
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  alignItems?: 'stretch' | 'flex-start' | 'flex-end' | 'center' | 'baseline';
  alignContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'stretch';
}

const Flex = (props: Flex): JSX.Element => {
  const { children } = props;

  return <StyledFlex {...props}>{children}</StyledFlex>;
};

export default Flex;

const StyledFlex = styled.div`
  display: flex;
  ${(props: Flex) => props.background && `background: ${props.background}`}
  ${(props: Flex) => props.direction && `flex-direction: ${props.direction}`}
  ${(props: Flex) => props.wrap && `flex-wrap: ${props.wrap}`}
  ${(props: Flex) =>
    props.justifyContent && `justify-content: ${props.justifyContent}`}
  ${(props: Flex) => props.alignItems && `align-items: ${props.alignItems}`}
  ${(props: Flex) =>
    props.alignContent && `align-content: ${props.alignContent}`}
`;
