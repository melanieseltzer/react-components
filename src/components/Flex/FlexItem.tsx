import React, { ReactNode } from 'react';
import styled from 'styled-components';

export interface FlexItemProps {
  className?: string;
  children: ReactNode;
  height?: string;
  width?: string;
  background?: string;
  flex?: string;
  order?: number;
  grow?: number;
  shrink?: number;
  basis?: string;
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'baseline'
    | 'stretch';
}

const FlexItem = (props: FlexItemProps) => {
  const { className, children } = props;

  return (
    <StyledFlexItem className={className} {...props}>
      {children}
    </StyledFlexItem>
  );
};

const StyledFlexItem = styled.div`
  ${(props: FlexItemProps) => props.flex && `flex: ${props.flex}`}
  ${(props: FlexItemProps) => props.height && `height: ${props.height}`}
  ${(props: FlexItemProps) => props.width && `width: ${props.width}`}
  ${(props: FlexItemProps) =>
    props.background && `background: ${props.background}`}
  ${(props: FlexItemProps) => props.order && `order: ${props.order}`}
  ${(props: FlexItemProps) => props.grow && `flex-grow: ${props.grow}`}
  ${(props: FlexItemProps) => props.shrink && `flex-shrink: ${props.shrink}`}
  ${(props: FlexItemProps) => props.basis && `flex-basis: ${props.basis}`}
  ${(props: FlexItemProps) =>
    props.alignSelf && `align-self: ${props.alignSelf}`}
`;

export default FlexItem;
