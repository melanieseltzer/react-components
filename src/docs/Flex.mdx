---
name: Flex
menu: Components
route: /flex
---

import { Playground, Props } from 'docz';
import Flex from '../components/Flex/Flex';
import FlexItem from '../components/FlexItem/FlexItem';

# Flex

An abstraction over a Flexbox ðŸš€ Defines a flex container (enables flex for all direct children).

## Props

<Props of={Flex} />

## Usage

<Playground>
  <Flex justifyContent="space-around">
    <FlexItem background="red">Hi</FlexItem>
    <FlexItem background="red">hi</FlexItem>
    <FlexItem background="red">hi</FlexItem>
  </Flex>
</Playground>
