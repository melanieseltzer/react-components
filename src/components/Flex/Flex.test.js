import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Flex from '.';

afterEach(cleanup);

describe('Flex component', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <Flex>
        <Flex.Item>Hi</Flex.Item>
      </Flex>
    );
    expect(container).toMatchSnapshot();
  });

  it('should be rendered with at least one child component', () => {
    const { container } = render(<Flex>Hi</Flex>);
    expect(container.firstChild.hasChildNodes).toBeTruthy();
  });

  it('should render with a class of test', () => {
    const { container } = render(<Flex className="test">Hi</Flex>);
    expect(container.firstChild.classList.contains('test')).toBe(true);
  });
});
