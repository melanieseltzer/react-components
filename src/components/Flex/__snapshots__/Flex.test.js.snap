// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex component matches the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
    >
      Hi
    </div>
  </div>
</div>
`;
